modules/vmfcore/src/statistics.cpp

```diff
@@ -0,0 +1,1276 @@
+                        break;
+                    case Variant::type_real:
+                        m_value = Variant( m_value.get_real() + fieldValue.get_real() );
+                        break;
+                    default:
+                        VMF_EXCEPTION( vmf::NotImplementedException, "Operation not applicable to this data type" );
+                    }
+                break;
+            case StatAction::Remove:
+                if( m_value.isEmpty() )
+                    VMF_EXCEPTION( vmf::NotImplementedException, "Operation not applicable to this data type" );
+                else if( m_value.getType() != fieldValue.getType() )
+                    VMF_EXCEPTION( vmf::TypeCastException, "Type mismatch" );
+                switch( m_value.getType() )
+                {
+                case Variant::type_integer:
+                    m_value = Variant( m_value.get_integer() - fieldValue.get_integer() );
+                    break;
+                case Variant::type_real:
+                    m_value = Variant( m_value.get_real() - fieldValue.get_real() );
+                    break;
+                default:
+                    VMF_EXCEPTION( vmf::NotImplementedException, "Operation not applicable to this data type" );
+                }
+                break;
+            }
+            return true;
+        }
+    virtual Variant value() const
+        {
+            std::unique_lock< std::mutex > lock( m_lock );
+            return m_value;
+        }
+
+private:
+    mutable std::mutex m_lock;
+    Variant m_value;
+
+public:
+    static StatOpBase* createInstance()
+        { return new StatOpSum(); }
+};
+
+class StatOpLast: public StatOpBase
+{
+public:
+    StatOpLast()
+        {}
+    virtual ~StatOpLast()
+        {}
+
+public:
+    virtual std::string name() const
+        { return StatOpFactory::builtinName( StatOpFactory::BuiltinOp::Last ); }
+    virtual void reset()
+        {
+            std::unique_lock< std::mutex > lock( m_lock );
+            m_value = Variant();
+        }
+    virtual bool handle( StatAction::Type action, const Variant& fieldValue )
+        {
+            std::unique_lock< std::mutex > lock( m_lock );
+            switch( action )
+            {
+            case StatAction::Add:
+                m_value = fieldValue;
+                break;
+            case StatAction::Remove:
+                return false;
+            }
+            return true;
+        }
+    virtual Variant value() const
+        {
+            std::unique_lock< std::mutex > lock( m_lock );
+            return m_value;
+        }
+
+private:
+    mutable std::mutex m_lock;
+    Variant m_value;
+
+public:
+    static StatOpBase* createInstance()
+        { return new StatOpLast(); }
+};
+
+// class StatOpFactory
+
+StatOpBase* StatOpFactory::create( const std::string& name )
+{
+    // TODO: use lock for thread-safe access to UserOpMap instance
```

Comment:
good point!
it looks simple if we use `StatOpFactory::m_lock` to make this class thread-safe.